pkg/util/connectionpool/grpc/pool.go

```diff
@@ -0,0 +1,301 @@
+
+	defaultDialOpts = []grpc.DialOption{grpc.WithBlock(),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithCodec(GetCodecInstance())}
+	_ connectionpool.Pool = (*Pool)(nil)
+)
+
+const (
+	closed uint32 = iota
+	normal
+)
+
+type (
+	// Spec describe Pool
+	Spec struct {
+		BorrowTimeout      time.Duration
+		ConnectTimeout     time.Duration
+		ConnectionsPerHost int
+		DialOptions        []grpc.DialOption
+	}
+
+	// Pool is the grpc client pool
+	Pool struct {
+		segment  sync.Map
+		factory  connectionpool.CreateConnFactory
+		spec     *Spec
+		mu       sync.RWMutex
+		isClosed bool
+	}
+
+	segment struct {
+		count   int32
+		clients chan *ClientConn
+		pool    *Pool
+	}
+
+	// ClientConn is the wrapper for a grpc client conn
+	ClientConn struct {
+		*grpc.ClientConn
+		segment  *segment
+		returned uint32
+	}
+)
+
+func newPool(factory connectionpool.CreateConnFactory, spec *Spec) (*Pool, error) {
+	// grpc-go office suggest use one Connection and multiple stream.
+	// and think of to balance pressure of tcp packet parsing on multiple cpus.
+	// refer to https://github.com/grpc/grpc-go/issues/3005, so default we design
+	// init conn num is 2. we adopt producer-consumer model, for each addr,
+	// first call Get() and we create init num conn to pool , then if consumer
+	// consume too fast, we would create new conn until reach Spec.MaxConnectionsPerHost
+	// if reach max, we use conn with random policy
+	if spec.ConnectionsPerHost <= 0 {
+		spec.ConnectionsPerHost = 2
+	}
+	if spec.BorrowTimeout == 0 {
+		spec.BorrowTimeout = 500 * time.Millisecond
+	}
+	if spec.ConnectTimeout == 0 {
+		spec.ConnectTimeout = 200 * time.Millisecond
+	}
+
+	if len(spec.DialOptions) == 0 {
+		spec.DialOptions = defaultDialOpts
+	}
+
+	p := &Pool{
+		factory: factory,
+		spec:    spec,
+	}
+	return p, nil
+}
+
+// New creates a new clients pool with the given initial and maximum capacity.
+// Returns an error if the initial clients could not be created
+func New(spec *Spec) (*Pool, error) {
+	return newPool(nil, spec)
+}
+
+// NewWithFactory creates a new clients pool with the given initial and maximum
+// capacity. The context parameter would be passed to the factory method during initialization.
+// Returns an error if the initial clients could not be created.
+func NewWithFactory(factory connectionpool.CreateConnFactory, spec *Spec) (*Pool, error) {
+	if factory == nil {
+		return nil, ErrorParams
+	}
+	return newPool(factory, spec)
+}
+
+// Close empties the pool calling Close on all its clients.
+// You can call Close while there are outstanding clients.
+// The pool channel is then closed, and Get will not be allowed anymore
+func (p *Pool) Close() {
+	p.mu.RLock()
+	if p.isClosed {
+		p.mu.RUnlock()
+		return
+	}
+	p.mu.RUnlock()
+	p.mu.Lock()
+	defer func() {
+		p.isClosed = true
+		p.mu.Unlock()
+	}()
+
+	p.segment.Range(func(key, value any) bool {
+		close(value.(*segment).clients)
```

Comment:
it is possible that close a closed channel, and channel should be closed at the producer side.